src/openfl/text/_internal/ShapeCache.hx

```diff
@@ -0,0 +1,109 @@
+package openfl.text._internal;
+import haxe.ds.IntMap;
+import haxe.ds.StringMap;
+
+#if !openfl_debug
+@:fileXml('tags="haxe,release"')
+@:noDebug
+#end
+
+@:access(openfl.text.TextFormat)
+@SuppressWarnings("checkstyle:FieldDocComment")
+class ShapeCache
+{
+	private var __shortWordMap:StringMap<StringMap<Array<GlyphPosition>>>;
+	private var __longWordMap:StringMap<IntMap<CacheMeasurement>>;
+
+	public function new()
+	{
+		__shortWordMap = new StringMap();
+		__longWordMap = new StringMap();
+	}
+
+	private static function hashFunction(key:String):Int
+	{
+		var hash = 0, i, chr;
+		for (i in 0...key.length)
+		{
+			chr = key.charCodeAt(i);
+			hash = ((hash << 5) - hash) + chr;
+			hash |= 0;
+		}
+		return hash;
+	}
+
+	public function cache(formatRange:TextFormatRange, getPositions:#if (js && html5) ()->, wordKey:String = null #else TextLayout #end):Dynamic
+	{
+		var formatKey:String = formatRange.format.__cacheKey;
+		#if (!js && !html5)
+		var wordKey:String = getPositions.text;
+		#end
+		if (wordKey.length > 15)
+		{
+			return __cacheLongWord(wordKey, formatKey, getPositions);
+		}
+		else {
+			return __cacheShortWord(wordKey, formatKey, getPositions);
+		}
+		
+	}
+
+	private function __cacheShortWord(wordKey:String, formatKey:String, getPositions:#if (js && html5) ()->):Array<Float #else TextLayout):Array<GlyphPosition #end>
```

Comment:
I would be slightly careful of passing a function in. I don't remember if js has issues with closures, but if so that object generation could add up. May not have an effect if you never actually call the function, which is the case if it's in the cache already

Reply:
Maybe there is a better way to unify it so I don't end up with separate methods for native and js (or a lot more code that will increase complexity)...I think it might be ok for now. If we could rework getPositions for js && html5, it would eliminate the need.